### PR TITLE
GenericAlias pickle support

### PR DESCRIPTION
**Pull request by emmatyping**
So I basically copied the implementation in [`range_reduce`](https://github.com/python/cpython/blob/46874c26ee1fc752e2e6930efa1d223b2351edb8/Objects/rangeobject.c#L609), but it is failing the test for some reason with the following traceback:

```
Traceback (most recent call last):
  File "C:\Users\ethanhs\cpython\Lib\test\test_genericalias.py", line 179, in test_pickle
    s = pickle.dumps(alias)
TypeError: object.__reduce_ex__() takes exactly one argument (0 given)
```
I'm not really sure why this would happen, and my efforts to fix it haven't worked.

### CODE REVIEW

File: Objects/descrobject.c [hunk 1]
```diff
@@ -2077,10 +2077,18 @@ ga_subclasscheck(PyObject *self, PyObject *Py_UNUSED(ignored))
...
+{
+    return Py_BuildValue("(O(OOO))", Py_TYPE(self),
+                         self->origin, self->args, self->parameters);
+}
+
 static PyMethodDef ga_methods[] = {
     {"__mro_entries__", ga_mro_entries, METH_O},
     {"__instancecheck__", ga_instancecheck, METH_O},
     {"__subclasscheck__", ga_subclasscheck, METH_O},
+    {"__reduce__", ga_reduce, METH_NOARGS},
```
Comment (emmatyping) [inline Objects/descrobject.c:2133]: Various things in CPython have `METH_O`, `METH_VARARGS`, and `METH_NOARGS` here for their `__reduce__`, so I'm not certain which is the best.

Comment (gvanrossum) [reply Objects/descrobject.c:2133]: Shouldn't matter -- this just declares how to call the C function.

**Comment by gvanrossum** [comment]
I think the crash is because the pickle module first looks up `__reduce_ex__` and then `__reduce__`. Apparently it finds a hit for the former -- there seems to be an `object.__reduce_ex__`. You probably also need to add something to the list of getattr exceptions.

File: Lib/test/test_genericalias.py [hunk 2]
```diff
@@ -1,14 +1,18 @@
...
     defaultdict, deque, OrderedDict, Counter, UserDict, UserList
 )
 from collections.abc import *
 from contextlib import AbstractContextManager, AbstractAsyncContextManager
 from io import IOBase
 from re import Pattern, Match
+from types import GenericAlias
 
+from typing import TypeVar
+T = TypeVar('T')
```
Comment (emmatyping) [inline Lib/test/test_genericalias.py:15]: I had to put this in the global scope otherwise it would fail pickling :(

File: Objects/descrobject.c [hunk 3]
```diff
@@ -2112,7 +2142,7 @@ ga_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 // - __eq__
 PyTypeObject Py_GenericAliasType = {
     PyVarObject_HEAD_INIT(&PyType_Type, 0)
-    .tp_name = "GenericAlias",
+    .tp_name = "types.GenericAlias",
```
Comment (emmatyping) [inline Objects/descrobject.c:2166]: Hopefully this is okay, pickle would try to import `GenericAlias` from builtins otherwise.

**Comment by emmatyping** [comment]
>You probably also need to add something to the list of getattr exceptions.

Ah! Of course, I should have thought of that. Thank you this should now work as expected.

File: Objects/descrobject.c (hunk 3 above)
Comment (gvanrossum) [reply Objects/descrobject.c:2166]: Yes, indeed this seems how it's done, see e.g. odictobject.c:1550, picklebufobject.c:208, namespaceobject.c:206 (all in Objects).

File: Lib/test/test_genericalias.py (hunk 2 above)
Comment (gvanrossum) [reply Lib/test/test_genericalias.py:15]: Yeah, that's fine. I like this better than putting `global T` in the test function. :-)

File: Objects/descrobject.c [hunk 4]
```diff
@@ -1999,6 +1999,9 @@ static const char* const attr_exceptions[] = {
     "__args__",
     "__parameters__",
     "__mro_entries__",
+    "__reduce_ex__",
```
Comment (gvanrossum) [inline Objects/descrobject.c]: Explain in a comment why you need this?

**Review by gvanrossum — APPROVED** [review]
I'd prefer to recompute `parameters` in `__setstate__` rather than making it part of the pickled data; at some point I'd like to compute it lazily. It's a pure function of `args` anyway.

**Review by gvanrossum — COMMENTED** [review]
Sorry, meant not to approve yet.

**Review by gvanrossum — CHANGES_REQUESTED** [review]
(try again)

**Comment by emmatyping** [comment]
Done! I had to move `make_parameters` above `ga_setstate`, but I kind of like that anyway since it is with the "utility functions".

**Review by gvanrossum — APPROVED** [review]

**Pull request merged**